harness/determined/common/api/authentication.py

```diff
@@ -22,6 +23,37 @@ def salt_and_hash(password: str) -> str:
         return password
 
 
+def check_password_complexity(password: Optional[str]) -> None:
+    """Raises a ValueError if the password does not meet complexity requirements.
+
+    The complexity requirements are:
+        - Must be at least 8 characters long.
+        - Must contain at least one upper-case letter.
+        - Must contain at least one lower-case letter.
+        - Must contain at least one number.
+
+    Args:
+        password: a password to check.
+
+    Raises:
+        ValueError: an error describing why the password does not meet complexity requirements.
+    """
+    if password is None:
```

Comment:
Prefer `if not password`. This is a good exception to raise if `password == ""`.

Reply:
What I'd like to do here is raise an error that's as useful as possible so the user doesn't have to guess a bunch of times* at what the password requirements are.

*of course, ideally this is also going to be documented in useful places, but the CLI should help too if it can.

Worst case is something like:

> ""

`password cannot be blank`

> "Abc123!"

`password must have at least 8 characters`

> "atleast8characters"

`password must include an uppercase letter`

> "ATLEAST8CHARACTERS"

`password must include a lowercase letter`

> "atLeastEightCharacters"

`password must include a number`

> "Ih8this"

`password must have at least 8 characters`

Maybe a better approach would be to catch as many errors as possible, collect them, and raise an error that contains all the problems. Or maybe there should just be a standard error string that points the user to the password requirements in documentation or summarizes all the requirements together somehow.

Reply:
That's a nice feature.

Now that we're not early exiting, somewhere the system again needs to handle `None` as password. I still think it belongs in this function rather than in callers'. I can't think of any slick ways of doing it outside of adding a special case checking `None` and reassigning the empty string or something.

It's a nit and do whatever you like, but I think it could help readability to rename `good` and `bad` to something that describes the character rather than how the character will be used. `check` is easy, but `x` isn't a great name. So I don't know, maybe something better comes to you.

Reply:
>  the system again needs to handle None as password.
I like it!